### ISSUE TITLE
share weights

### ISSUE THREAD

**Issue by nnop** [nnop/notes#126]
- Weight sharing #500 ->  #546
- Improve / Fix Weight Sharing #1211
- learnable_param_ids_ fix #2909
- Fix weight sharing #2866

**Comment by nnop** [comment]
**snapshot problem**

- Snapshot model weights/solver state to HDF5 files #2836
- Caffemodel snapshots with shared weights don't have multiple copies #2946
- Caffemodel protobuf snapshots with shared weights shouldn't save multiple copies #2955

**Comment by nnop** [comment]
**BatchNorm layer share weights**
- questions about batch norm auto-upgrade.  caffe@5120
- Cannot explicitly "name" BatchNorm parameters for sharing (Siamese network) caffe@5171

Fixed by 
- Fixing harsh upgrade_proto for `"BatchNorm"` layer caffe@5184
